src/iso.c

```diff
@@ -0,0 +1,489 @@
+    if(extent_num < 0) {
+        return 0;
+    }
+    struct iso_dir *to_return = kmalloc(sizeof(*to_return));
+    struct iso_point *iso_p = iso_media_open(ata_unit);
+    iso_media_seek(iso_p, ATAPI_BLOCKSIZE * extent_num, SEEK_SET);
+    to_return->cur_offset = 0;
+    to_return->extent_offset = extent_num;
+    to_return->ata_unit = ata_unit;
+    to_return->data_length = dl;
+
+    iso_media_close(iso_p);
+    return to_return;
+}
+
+//Documentation in header
+struct directory_record *iso_dread(struct iso_dir *read_from) {
+    if(read_from->cur_offset + 1 >= read_from->data_length) {
+        //"return null" if at end of directory record
+        return 0;
+    }
+
+    struct iso_point *iso_p = iso_media_open(read_from->ata_unit);
+    iso_media_seek(iso_p, read_from->extent_offset * ATAPI_BLOCKSIZE + read_from->cur_offset, SEEK_SET);  // seek to current offset of iso_p
+    struct directory_record *next_dr = kmalloc(sizeof(*next_dr));
+
+    //if not the end of the directory record, fill the dest by dereferencing dest_ptr
+    int success = get_directory_record(iso_p, next_dr);
+    if (!success) {
+        iso_media_close(iso_p);
+        return 0;
+    }
+
+    read_from->cur_offset = iso_p->cur_offset;
+    //no need to reset extent, that shouldn't be an issue
+    iso_media_close(iso_p);
+    return next_dr;
+}
+
+//Documentation in header
+int iso_fclose(struct iso_file *file) {
+    if(file) {
+        kfree(file);
+        return 0;
+    }
+    return -1;
+}
+
+//Documentation in header
+struct iso_file *iso_fopen(const char *pname, int ata_unit) {
+    struct iso_file *file = kmalloc(sizeof(struct iso_file));
+    strcpy(file->pname, pname);
+    int dl;  //the data_length of the last item on the path
+    int file_offset = iso_look_up(pname, &dl, ata_unit);
+    if (file_offset < 0) {
+        console_printf("Failed to iso_open the file path \"%s\".\n", pname);
+        return 0;
+    }
+    file->cur_offset = 0;
+    file->extent_offset = file_offset;
+    file->ata_unit = ata_unit;
+    file->data_length = dl;
+
+    return file;
+}
+
+//Documentation in header
+int iso_fread(void *dest, int elem_size, int num_elem, struct iso_file *file) {
+    struct iso_point *iso_p = iso_media_open(file->ata_unit);
+    iso_media_seek(iso_p, ISO_BLOCKSIZE * file->extent_offset + file->cur_offset, SEEK_SET);
+
+    int at_EOF = 0;
+    int bytes_to_disk_read = elem_size * num_elem;
+    int bytes_to_file_read = bytes_to_disk_read;
+    if ((bytes_to_file_read + file->cur_offset) > file->data_length) {
+        bytes_to_file_read = file->data_length - file->cur_offset + 1;  // Don't actually want to read off the end of the file
+        bytes_to_disk_read = bytes_to_file_read - 1;
+        at_EOF = 1;
+    }
+    int bytes_read = iso_media_read(dest, 1, bytes_to_disk_read, iso_p);
+    if(bytes_read != bytes_to_disk_read) {
+        console_printf("file reading error\n");
+        iso_media_close(iso_p);
+        return -2;
```

Comment:
No, -1 is the implemented definition of EOF. -2 is reserved here for a file reading error. The positive integers and zero are reserved for the number of bytes successfully read.